README
NAME
    HTML::Restrict - Strip unwanted HTML tags and attributes

VERSION
    version 2.2.2

SYNOPSIS
        use HTML::Restrict;

        my $hr = HTML::Restrict->new();

        # use default rules to start with (strip away all HTML)
        my $processed = $hr->process('  <b>i am bold</b>  ');

        # $processed now equals: 'i am bold'

        # Now, a less restrictive example:
        use HTML::Restrict;

        my $hr = HTML::Restrict->new(
            rules => {
                b   => [],
                img => [qw( src alt / )]
            }
        );

        my $html = q[<body><b>hello</b> <img src="pic.jpg" alt="me" id="test" /></body>];
        my $processed = $hr->process( $html );

        # $processed now equals: <b>hello</b> <img src="pic.jpg" alt="me" />

DESCRIPTION
    This module uses HTML::Parser to strip HTML from text in a restrictive
    manner. By default all HTML is restricted. You may alter the default
    behaviour by supplying your own tag rules.

CONSTRUCTOR AND STARTUP
  new()
    Creates and returns a new HTML::Restrict object.

        my $hr = HTML::Restrict->new()

    HTML::Restrict doesn't require any params to be passed to new. If your
    goal is to remove all HTML from text, then no further setup is required.
    Just pass your text to the process() method and you're done:

        my $plain_text = $hr->process( $html );

    If you need to set up specific rules, have a look at the params which
    HTML::Restrict recognizes:

    *   "rules => \%rules"

        Sets the rules which will be used to process your data. By default
        all HTML tags are off limits. Use this argument to define the HTML
        elements and corresponding attributes you'd like to use.
        Essentially, consider the default behaviour to be:

            rules => {}

        Rules should be passed as a HASHREF of allowed tags. Each hash value
        should represent the allowed attributes for the listed tag. For
        example, if you want to allow a fair amount of HTML, you can try
        something like this:

            my %rules = (
                a       => [qw( href target )],
                b       => [],
                caption => [],
                center  => [],
                em      => [],
                i       => [],
                img     => [qw( alt border height width src style )],
                li      => [],
                ol      => [],
                p       => [qw(style)],
                span    => [qw(style)],
                strong  => [],
                sub     => [],
                sup     => [],
                table   => [qw( style border cellspacing cellpadding align )],
                tbody   => [],
                td      => [],
                tr      => [],
                u       => [],
                ul      => [],
            );

            my $hr = HTML::Restrict->new( rules => \%rules )

        Or, to allow only bolded text:

            my $hr = HTML::Restrict->new( rules => { b => [] } );

        Allow bolded text, images and some (but not all) image attributes:

            my %rules = (
                b   => [ ],
                img => [qw( src alt width height border / )
            );
            my $hr = HTML::Restrict->new( rules => \%rules );

        Since HTML::Parser treats a closing slash as an attribute, you'll
        need to add "/" to your list of allowed attributes if you'd like
        your tags to retain closing slashes. For example:

            my $hr = HTML::Restrict->new( rules =>{ hr => [] } );
            $hr->process( "<hr />"); # returns: <hr>

            my $hr = HTML::Restrict->new( rules =>{ hr => [qw( / )] } );
            $hr->process( "<hr />"); # returns: <hr />

        HTML::Restrict strips away any tags and attributes which are not
        explicitly allowed. It also rebuilds your explicitly allowed tags
        and places their attributes in the order in which they appear in
        your rules.

        So, if you define the following rules:

            my %rules = (
                ...
                img => [qw( src alt title width height id / )]
                ...
            );

        then your image tags will all be built like this:

            <img src=".." alt="..." title="..." width="..." height="..." id=".." />

        This gives you greater consistency in your tag layout. If you don't
        care about element order you don't need to pay any attention to
        this, but you should be aware that your elements are being
        reconstructed rather than just stripped down.

        As of 2.1.0, you can also specify a regex to be tested against the
        attribute value. This feature should be considered experimental for
        the time being:

            my $hr = HTML::Restrict->new(
                rules => {
                    iframe => [
                        qw( width height allowfullscreen ),
                        {   src         => qr{^http://www\.youtube\.com},
                            frameborder => qr{^(0|1)$},
                        }
                    ],
                    img => [ qw( alt ), { src => qr{^/my/images/} }, ],
                },
            );

            my $html = '<img src="http://www.example.com/image.jpg" alt="Alt Text">';
            my $processed = $hr->process( $html );

            # $processed now equals: <img alt="Alt Text">

    *   "trim => [0|1]"

        By default all leading and trailing spaces will be removed when text
        is processed. Set this value to 0 in order to disable this
        behaviour.

    *   "uri_schemes => [undef, 'http', 'https', 'irc', ... ]"

        As of version 1.0.3, URI scheme checking is performed on all href
        and src tag attributes. The following schemes are allowed out of the
        box. No action is required on your part:

            [ undef, 'http', 'https' ]

        (undef represents relative URIs). These restrictions have been put
        in place to prevent XSS in the form of:

            <a href="javascript:alert(document.cookie)">click for cookie!</a>

        See URI for more detailed info on scheme parsing. If, for example,
        you wanted to filter out every scheme barring SSL, you would do it
        like this:

            uri_schemes => ['https']

        This feature is new in 1.0.3. Previous to this, there was no schema
        checking at all. Moving forward, you'll need to whitelist explicitly
        all URI schemas which are not supported by default. This is in
        keeping with the whitelisting behaviour of this module and is also
        the safest possible approach. Keep in mind that changes to
        uri_schemes are not additive, so you'll need to include the defaults
        in any changes you make, should you wish to keep them:

            # defaults + irc + mailto
            uri_schemes => [ 'undef', 'http', 'https', 'irc', 'mailto' ]

    *   allow_declaration => [0|1]

        Set this value to true if you'd like to allow/preserve DOCTYPE
        declarations in your content. Useful when cleaning up your own
        static files or templates. This feature is off by default.

            my $html = q[<!doctype html><body>foo</body>];

            my $hr = HTML::Restrict->new( allow_declaration => 1 );
            $html = $hr->process( $html );
            # $html is now: "<!doctype html>foo"

    *   allow_comments => [0|1]

        Set this value to true if you'd like to allow/preserve HTML comments
        in your content. Useful when cleaning up your own static files or
        templates. This feature is off by default.

            my $html = q[<body><!-- comments! -->foo</body>];

            my $hr = HTML::Restrict->new( allow_comments => 1 );
            $html = $hr->process( $html );
            # $html is now: "<!-- comments! -->foo"

    *   replace_img => [0|1|CodeRef]

        Set the value to true if you'd like to have img tags replaced with
        "[IMAGE: ...]" containing the alt attribute text. If you set it to a
        code reference, you can provide your own replacement (which may even
        contain HTML).

            sub replacer {
                my ($tagname, $attr, $text) = @_; # from HTML::Parser
                return qq{<a href="$attr->{src}">IMAGE: $attr->{alt}</a>};
            }

            my $hr = HTML::Restrict->new( replace_img => \&replacer );

        This attribute will only take effect if the img tag is not included
        in the allowed HTML.

    *   strip_enclosed_content => [0|1]

        The default behaviour up to 1.0.4 was to preserve the content
        between script and style tags, even when the tags themselves were
        being deleted. So, you'd be left with a bunch of JavaScript or CSS,
        just with the enclosing tags missing. This is almost never what you
        want, so starting at 1.0.5 the default will be to remove any script
        or style info which is enclosed in these tags, unless they have
        specifically been whitelisted in the rules. This will be a sane
        default when cleaning up content submitted via a web form. However,
        if you're using HTML::Restrict to purge your own HTML you can be
        more restrictive.

            # strip the head section, in addition to JS and CSS
            my $html = '<html><head>...</head><body>...<script>JS here</script>foo';

            my $hr = HTML::Restrict->new(
                strip_enclosed_content => [ 'script', 'style', 'head' ]
            );

            $html = $hr->process( $html );
            # $html is now '<html><body>...foo';

        The caveat here is that HTML::Restrict will not try to fix broken
        HTML. In the above example, if you have any opening script, style or
        head tags which don't also include matching closing tags, all
        following content will be stripped away, regardless of any parent
        tags.

        Keep in mind that changes to strip_enclosed_content are not
        additive, so if you are adding additional tags you'll need to
        include the entire list of tags whose enclosed content you'd like to
        remove. This feature strips script and style tags by default.

SUBROUTINES/METHODS
  process( $html )
    This is the method which does the real work. It parses your data,
    removes any tags and attributes which are not specifically allowed and
    returns the resulting text. Requires and returns a SCALAR.

CAVEATS
    Please note that all tag and attribute names passed via the rules param
    must be supplied in lower case.

        # correct
        my $hr = HTML::Restrict->new( rules => { body => ['onload'] } );

        # throws a fatal error
        my $hr = HTML::Restrict->new( rules => { Body => ['onLoad'] } );

MOTIVATION
    There are already several modules on the CPAN which accomplish much of
    the same thing, but after doing a lot of poking around, I was unable to
    find a solution with a simple setup which I was happy with.

    The most common use case might be stripping HTML from user submitted
    data completely or allowing just a few tags and attributes to be
    displayed. With the exception of URI scheme checking, this module
    doesn't do any validation on the actual content of the tags or
    attributes. If this is a requirement, you can either mess with the
    parser object, post-process the text yourself or have a look at one of
    the more feature-rich modules in the SEE ALSO section below.

    My aim here is to keep things easy and, hopefully, cover a lot of the
    less complex use cases with just a few lines of code and some brief
    documentation. The idea is to be up and running quickly.

SEE ALSO
    HTML::TagFilter, HTML::Defang, HTML::Declaw, HTML::StripScripts,
    HTML::Detoxifier, HTML::Sanitizer, HTML::Scrubber

ACKNOWLEDGEMENTS
    Thanks to Raybec Communications <http://www.raybec.com> for funding my
    work on this module and for releasing it to the world.

    Thanks also to the following for patches, bug reports and assistance:

    Mark Jubenville (ioncache)

    Duncan Forsyth

    Rick Moore

    Arthur Axel 'fREW' Schmidt

    perlpong

    David Golden

    Graham TerMarsch

    Dagfinn Ilmari Mannsåker

    Graham Knop

    Carwyn Ellis

AUTHOR
    Olaf Alders <olaf@wundercounter.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Olaf Alders.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for HTML-Restrict

{{$NEXT}}

v3.0.2    2023-04-23 17:06:28Z
        - Add create_newlines to replace BR and P by 1, resp. 2 newlines
          (GH#46) (E. Choroba)

v3.0.1    2022-09-22 22:27:40Z
        - Add a filter_text attribute. This can be used to disable text
          encoding. (GH#44) (ccottingham-presinet)

v3.0.0    2019-03-11 13:52:08Z
        - Better fix for handling malformed tags. Removes max_parser_loops(),
          which was introduced in 2.4.0 (GH#37) (Graham Knop)

v2.5.0    2019-02-08 22:18:11Z
        - Strip some control characters from links (GH#34) (Olaf Alders)
        - Enable empty_element_tags in HTML::Parser (GH#35) (Olaf Alders)

v2.4.1    2019-02-05 14:13:16Z
        - Bump version of Type::Tiny to 1.002001. (GH#33) (Olaf Alders).  Issue
          reported by Slaven Rezić (GH#32).

v2.4.0    2019-02-05 02:51:05Z
        - Process text until it returns the same value twice. (GH#31) (Olaf
          Alders).  Issue raised in (GH#29) by Juraj Major.
        - Add max_parser_loops attribute

v2.3.0    2018-02-09 16:00:41Z
        - Add code filters for attribute values (GH#27) (Graham Knop)

2.2.4     2017-03-01 14:22:13-05:00 America/Toronto
        - Updates SEE ALSO pod section (Ben Bullock)

2.2.3     2016-07-26 18:00:55-04:00 America/Toronto
        - Now with fewer non-core prerequisites (Karen Etheridge)

2.2.2     2014-10-16 22:16:29-04:00 America/Toronto
        - Fixes attribute value encoding. (Graham Knop)

2.2.1     2014-08-18 17:22:05EDT-0400 America/Toronto
        - Replace MooX::Types::MooseLike with Type::Tiny. (Samuel Kaufman)

2.2.0   2014-08-15 23:01:22EDT-0400 America/Toronto
        - Fix Changes file
        - Bump to next minor version rather than changing the format of the
          version numbers

2.1.9   2013-12-10
        - Swaps namespace::clean for namespace::autoclean. (Fitz Elliott)

2.1.8   2013-06-06
        - Unlike 2.1.7, This release doesn't contain a local::lib.

2.1.7   2013-06-05
        - Removes Perl6::Junction in order to squash smartmatch warnings.

2.1.6   2013-05-06
        - Fixes incorrect use of Moo's weak_ref. (Thanks to Carwyn Ellis,
          Dagfinn Ilmari Mannsåker and Graham Knop for help with this.)

2.1.5   2013-05-01
        - Eliminates a circular reference. (Graham TerMarsch)

2.1.4   2013-04-19
        - Minor version should have been bumped in previous release. To bump it
          now would be confusing.
        - Require Moo 1.001000 as 1.000008 breaks this module.

2.1.3   2013-04-18
        - Add replace_img feature (David Golden).
        - Add 'cite' to list of attributes for URI scheme checks (David
          Golden).

2.1.2   2013-03-20
        - Fixes broken Pod in example script.
        - Adds x_contributors to META.json.

2.1.1   2013-03-08
        - Updates Pod.

2.1.0   2013-03-02
        - Allows attributes to be validated against regexes (perlpong).

2.0.0   2013-02-27
        - Bumps version to 2.0.0 due to new features breaking backwards
          compatibility.
        - Forces all tag rules to be provided in lower case.  Breaks
          backcompat, but mixed or upper case rules would have been silently
          ignored.  So, this is really a bug fix.
        - Strips content between script tags by default. Breaks backcompat,
          but the default behaviour was likely not what anyone wanted.
        - Strips content between style tags by default.  Breaks backcompat,
          but the default behaviour was likely not what anyone wanted.
        - Adds Pod for previously undocumented allow_comments,
          allow_declaration and strip_enclosed_content.
        - Ensures stripper stack is always reset (perlpong).
        - Protects against undef string comparison (perlpong).
        - Checks URI scheme for src attributes (perlpong).
        - Removes accessor/mutator documentation.  Would prefer if all params
          were set at object instatiation.
        - Adds ridiculously basic example script.


1.0.4   2012-05-22
        - Explicity require MooX::Types::MooseLike rather than
          MooX::Types::MooseLike::Base (Matt Phillips/Github #2)

1.0.3   2012-03-27
        - Replaces Moose with Moo for speed in non-persistent apps (Arthur Axel
          'fREW' Schmidt)
        - Adds uri scheme checking for href and src attributes
        - Adds (undocumented) allow_comments() and allow_declartion() methods

1.0.2   2011-10-11
        - No longer emit warnings when pure HTML is supplied for processing
            Rick Moore (RT #71553)

1.0.1   2011-06-30
        - Pod updates

1.0.0   2011-06-30
        - Bumped version to 1.0.0 This is no longer beta software.

0.06    2010-03-22
        - Fixed RT #55775 "Strings equating to boolean false returned as undef"
            (Duncan Forsyth)

0.05    2010-02-03
        - process( undef ) returns without attempting to do anything.  Avoids
            warnings about uninitialized vars.

0.04    2009-11-19
        - MooseX::Params::Validate 0.12 now required.  Some earlier versions do
            not include pos_validated_list()

0.03    2009-11-17
        - process() no longer dies when passed an undef -- patch provided by
            Mark Jubenville (ioncache)

0.02    2009-09-17
        - Improved documentation layout and clarity
            Removed documentation for parser() method -- seems like a bad idea

0.01    2009-09-17
        - Beta release


t/lib/HTML/Restrict/Rules/Test/Iframe.pm

use strict;
use 5.006;

package HTML::Restrict::Rules::Test::Iframe;

sub rules {
    +{
        iframe => [
            qw( width height ),
            {
                src         => qr{^http://www\.youtube\.com},
                frameborder => qr{^(0|1)$},
            }
        ],
    }
}

1;



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'HTML::Restrict',
    license             => 'perl',
    dist_abstract       => 'Strip unwanted HTML tags and attributes',
    dist_author         => q{Olaf Alders <olaf@wundercounter.com>},
    dist_version_from   => 'lib/HTML/Restrict.pm',
    build_requires => {
        'Data::Dump'                => 0,
        'HTML::Parser'              => 0,
        'Moose'                     => 0,
        'MooseX::Params::Validate'  => '0.12',
        'Perl6::Junction'           => 0,
        'Scalar::Util'              => 0,
        'Test::More'                => 0,
    },
    add_to_cleanup      => [ 'HTML-Restrict-*' ],
    create_makefile_pl => 'traditional',
);

$builder->create_build_script();


lib/HTML/Restrict/Constraints.pm

use strict;
use 5.006;

package HTML::Restrict::Constraints;

use Type::Library qw[ -base -declare Plugin Rules ];
use Type::Utils qw[ -all ];
use Types::Standard qw[ -types ];

use mro;

use namespace::clean;

my %load_cache;
my %plugin_cache;

sub looks_like_rules_plugin {
    my ($plugin) = @_;

    unless (exists $load_cache{$plugin}) {
        $load_cache{$plugin} = undef;

        return unless eval "require $plugin; 1";
        return unless $plugin->can ('rules');

        $load_cache{$plugin} = $plugin;
    };

    return $load_cache{$plugin}
}

sub rules_plugin_package {
    my ($class, $plugin) = @_;

    return unless defined $class;
    return unless defined $plugin;

    my $key = "$class/$plugin";

    unless (exists $plugin_cache{$key}) {
        for my $candidate ( @{ mro::get_linear_isa ($class) } ) {
            last if $plugin_cache{$key} = looks_like_rules_plugin ("${candidate}::$plugin");
        }

        $plugin_cache{$key} = looks_like_rules_plugin ($plugin)
            unless defined $plugin_cache{$key};
    }

    return $plugin_cache{$key};
}


declare Plugin,
    as Str,
    where { defined rules_plugin_package ('HTML::Restrict::Rules', $_) },
    message { "$_ doesn't look like rules plugin name" },
    ;

declare Rules,
    as HashRef,
    ;

coerce Rules,
    from Plugin,
    via { rules_plugin_package ('HTML::Restrict::Rules', $_)->rules },
    ;

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.030.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Strip unwanted HTML tags and attributes",
  "AUTHOR" => "Olaf Alders <olaf\@wundercounter.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "HTML-Restrict",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "HTML::Restrict",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dump" => 0,
    "HTML::Entities" => 0,
    "HTML::Parser" => 0,
    "List::Util" => "1.33",
    "Moo" => "1.002000",
    "Scalar::Util" => 0,
    "Sub::Quote" => 0,
    "Type::Tiny" => "1.002001",
    "Types::Standard" => "1.000001",
    "URI" => 0,
    "namespace::clean" => 0,
    "strict" => 0,
    "version" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.96",
    "warnings" => 0
  },
  "VERSION" => "v3.0.3",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Dump" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "HTML::Entities" => 0,
  "HTML::Parser" => 0,
  "List::Util" => "1.33",
  "Moo" => "1.002000",
  "Scalar::Util" => 0,
  "Sub::Quote" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "0.96",
  "Type::Tiny" => "1.002001",
  "Types::Standard" => "1.000001",
  "URI" => 0,
  "namespace::clean" => 0,
  "strict" => 0,
  "version" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
